### PR TITLE
Add grid partitioning option for PatchSidesetGenerator

### LINKED ISSUE #15829: Add grid partitioning option to PatchSidesetGenerator
## Reason

`PatchSidesetGenerator` is used for splitting sidesets up for radiative transfer. For rectangular faces,
splitting according to a regular grid seems to be the most reasonable.

## Design

Added capability implemented as branching in partitioning state. 

## Impact

new capability.

### LINKED ISSUE #15903: Add += operator to MooseEnum
## Reason
It would be useful to have a `+=` operator in `MooseEnum` that adds the items from another enum. This is useful in the case where a base class provides an enum and a derived class wants to  use the same enum but add options.

## Design
Add the operator that creates public access to `addEnumerationItem()`

## Impact
Opens up `MooseEnum` a little more. This isn't part of the original design, but I feel it's relevant and necessary.

### PR DESCRIPTION
**Pull request by snschune**
Adds the option to partition sidesets into patches using a regular grid. 
Some tests for that. Also contains provisions for cases where patches 
wind up empty. 
closes #15829
closes #15903

### CODE REVIEW

**Comment by loganharbour** [comment]
So what's goin on here?

**Comment by snschune** [comment]
```
ERROR: The library has been built without
Space Filling Curve support.  Using a linear
```

But that shouldn't happen. The partitioner is set to centroid and the screen output suggests that
the test runs just fine.

**Review by loganharbour — DISMISSED** [review]

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -136,11 +136,9 @@ MooseMesh::validParams()
                              "In particular you must supply this for GMSH meshes. "
                              "Note: This is completely ignored for ExodusII meshes!");
 
-  MooseEnum partitioning("default=-3 metis=-2 parmetis=-1 linear=0 centroid hilbert_sfc morton_sfc",
-                         "default");
   params.addParam<MooseEnum>(
       "partitioner",
-      partitioning,
+      partitioning(),
```
Comment (aeslaughter) [inline framework/src/mesh/MooseMesh.C:141]: This isn't necessary and I think makes reading the validParams method less clear, see my comment below in PatchSide...

File: modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C [hunk 2]
```diff
@@ -49,12 +50,13 @@ PatchSidesetGenerator::validParams()
...
       "n_patches", "n_patches>0", "Number of patches");
 
-  MooseEnum partitioning("default=-3 metis=-2 parmetis=-1 linear=0 centroid hilbert_sfc morton_sfc",
-                         "default");
+  MooseEnum partitioning = MooseMesh::partitioning(); // default MOOSE partitioning
+  partitioning += "grid";                             // ...but also add our own
   params.addParam<MooseEnum>(
       "partitioner",
       partitioning,
       "Specifies a mesh partitioner to use when splitting the mesh for a parallel computation.");
```
Comment (aeslaughter) [inline modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C:58]: params.set<MooseEnum>("partitioner") += "grid";

File: modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C [hunk 3]
```diff
@@ -188,8 +193,23 @@ PatchSidesetGenerator::generate()
...
+  else
+  {
+    // the enum of this class and MooseMesh enum are different
+    // so we have to go through these intermediate steps to not
+    // mess things up with partitioner
+    std::string pstring = _partitioner_name;
+    MooseEnum penum = MooseMesh::partitioning();
+    penum = pstring;
+    MooseMesh::setPartitioner(*boundary_mesh, penum, false, _pars, *this);
+    boundary_mesh->partition(_n_patches);
```
Comment (aeslaughter) [inline modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C:208]: I don't think you need the initial copy
```suggestion
    MooseEnum penum = MooseMesh::partitioning();
    penum = _partitioner_name;
    MooseMesh::setPartitioner(*boundary_mesh, penum, false, _pars, *this);
    boundary_mesh->partition(_n_patches);
```

File: modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C [hunk 4]
```diff
@@ -230,6 +250,97 @@ PatchSidesetGenerator::generate()
...
+      {
+        _console << "Note: For creating radiation patches for boundary " << _sideset
+                 << " using grid partitioner number of patches was changed from " << _n_patches
+                 << " to " << nx * ny << std::endl;
+        _n_patches = nx * ny;
+      }
+    }
+
+    Real dx = delta(0) / nx;
+    Real dy = delta(1) / ny;
```
Comment (aeslaughter) [inline modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C:285]: I know these are just temporaries, but the 'const' indicates that they don't change later. I just like that from a readability standpoint, you can take it or leave it. Just my 2 cents. There are a few others you could probably do this with if you want.
```suggestion
   const Real dx = delta(0) / nx;
   const Real dy = delta(1) / ny;
```

File: modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C [hunk 5]
```diff
@@ -230,6 +250,97 @@ PatchSidesetGenerator::generate()
...
+
+  // create a vector and sort it
+  std::vector<processor_id_type> processor_ids_vec;
+  for (auto & p : processor_ids)
+    processor_ids_vec.push_back(p);
+  std::sort(processor_ids_vec.begin(), processor_ids_vec.end());
+
+  // now remap the processor ids
+  std::map<processor_id_type, processor_id_type> processor_id_remap;
+  for (unsigned int j = 0; j < processor_ids_vec.size(); ++j)
```
Comment (aeslaughter) [inline modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C:331]:
```suggestion
  for (std::size_t j = 0; j < processor_ids_vec.size(); ++j)
```

**Review by aeslaughter — CHANGES_REQUESTED** [review]

File: framework/src/mesh/MooseMesh.C (hunk 1 above)
Comment (loganharbour) [reply framework/src/mesh/MooseMesh.C:141]: To make this clear, you suggest
```
params.addParam<MooseEnum>("partitioner", MooseMesh::partitioning(), "...");
params.set<MooseEnum>("partitioner") += "grid";
```

Is that correct?

Comment (aeslaughter) [reply framework/src/mesh/MooseMesh.C:141]: You don't need to add it, it is already there from the base class call

Comment (loganharbour) [reply framework/src/mesh/MooseMesh.C:141]: No it's not? `MeshGenerator != MooseMesh`

Comment (aeslaughter) [reply framework/src/mesh/MooseMesh.C:141]: Ah wait, I see why you did this. I was thinking that PatchSidesetGenerator inherited from this class, but it does not.

Comment (loganharbour) [reply framework/src/mesh/MooseMesh.C:141]: Alright - is this acceptable, then?

Comment (aeslaughter) [reply framework/src/mesh/MooseMesh.C:141]: yes

**Review by aeslaughter — APPROVED** [review]

**Pull request merged**